ext/semian/sliding_window.c

```diff
@@ -409,24 +418,30 @@ semian_simple_sliding_window_reject(VALUE self)
 
   sem_meta_lock(res->sem_id);
   {
-    // Store these values because we're going to be modifying the buffer.
-    int start = res->shmem->start;
-    int length = res->shmem->length;
-    dprintf("reject! - start:%d end:%d length:%d max_size:%d", res->shmem->start, res->shmem->end, res->shmem->length, res->shmem->max_size);
-
-    int cleared = 0;
-    for (int i = 0; i < length; ++i) {
-      int index = (start + i) % length;
-      int value = res->shmem->data[index];
-      VALUE y = rb_yield(RB_INT2NUM(value));
-      if (RTEST(y)) {
-        if (cleared++ != i) {
-          sem_meta_unlock(res->sem_id);
-          rb_raise(rb_eArgError, "reject! must delete monotonically");
+    semian_simple_sliding_window_shared_t *window = res->shmem;
+    const int start = window->start;
+    const int length = window->length;
+    const int max_size = window->max_size;
+
+    if (max_size && length) {
+      int wptr = (start + length + max_size - 1) % max_size;
+
+      dprintf("Before reject! start:%d length:%d max_size:%d", window->start, window->length, window->max_size);
+      for (int i = 0; i < length; ++i) {
+        const int rptr = (start + length + max_size - i - 1) % max_size;
+
+        const int value = window->data[rptr];
+        if (RTEST(rb_yield(RB_INT2NUM(value)))) {
+          window->length--;
+          window->data[wptr] = value;
+        } else {
+          window->data[wptr] = value;
+          wptr = (wptr + max_size - 1) % max_size;
```

Comment:
I really love how Python chose to make the default implementation for `%` be the same as the mathematical definition, which then allows you to express this exactly the way you'd like: `(wptr - 1) % max_size`